From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sat, 10 Feb 2024 14:44:02 +0100
Subject: [PATCH] Read packs - Content - Blocks - Map color from resource
 location

Adds a mapping from resource locations to `MapColor` instances.

This mapping is used to parse map colors from strings in the "Read packs - Content - Blocks - JSON parser" patch.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/world/level/material/MapColor.java b/src/main/java/net/minecraft/world/level/material/MapColor.java
index de73b38b78fee883c49a52e320b7a5d8ffbb9ee4..e5fd556fb5b2e1c59ed0405c2af2615b09b3ae2d 100644
--- a/src/main/java/net/minecraft/world/level/material/MapColor.java
+++ b/src/main/java/net/minecraft/world/level/material/MapColor.java
@@ -4,72 +4,78 @@ import com.google.common.base.Preconditions;
 
 public class MapColor {
     public static final MapColor[] MATERIAL_COLORS = new MapColor[64];
-    public static final MapColor NONE = new MapColor(0, 0);
-    public static final MapColor GRASS = new MapColor(1, 8368696);
-    public static final MapColor SAND = new MapColor(2, 16247203);
-    public static final MapColor WOOL = new MapColor(3, 13092807);
-    public static final MapColor FIRE = new MapColor(4, 16711680);
-    public static final MapColor ICE = new MapColor(5, 10526975);
-    public static final MapColor METAL = new MapColor(6, 10987431);
-    public static final MapColor PLANT = new MapColor(7, 31744);
-    public static final MapColor SNOW = new MapColor(8, 16777215);
-    public static final MapColor CLAY = new MapColor(9, 10791096);
-    public static final MapColor DIRT = new MapColor(10, 9923917);
-    public static final MapColor STONE = new MapColor(11, 7368816);
-    public static final MapColor WATER = new MapColor(12, 4210943);
-    public static final MapColor WOOD = new MapColor(13, 9402184);
-    public static final MapColor QUARTZ = new MapColor(14, 16776437);
-    public static final MapColor COLOR_ORANGE = new MapColor(15, 14188339);
-    public static final MapColor COLOR_MAGENTA = new MapColor(16, 11685080);
-    public static final MapColor COLOR_LIGHT_BLUE = new MapColor(17, 6724056);
-    public static final MapColor COLOR_YELLOW = new MapColor(18, 15066419);
-    public static final MapColor COLOR_LIGHT_GREEN = new MapColor(19, 8375321);
-    public static final MapColor COLOR_PINK = new MapColor(20, 15892389);
-    public static final MapColor COLOR_GRAY = new MapColor(21, 5000268);
-    public static final MapColor COLOR_LIGHT_GRAY = new MapColor(22, 10066329);
-    public static final MapColor COLOR_CYAN = new MapColor(23, 5013401);
-    public static final MapColor COLOR_PURPLE = new MapColor(24, 8339378);
-    public static final MapColor COLOR_BLUE = new MapColor(25, 3361970);
-    public static final MapColor COLOR_BROWN = new MapColor(26, 6704179);
-    public static final MapColor COLOR_GREEN = new MapColor(27, 6717235);
-    public static final MapColor COLOR_RED = new MapColor(28, 10040115);
-    public static final MapColor COLOR_BLACK = new MapColor(29, 1644825);
-    public static final MapColor GOLD = new MapColor(30, 16445005);
-    public static final MapColor DIAMOND = new MapColor(31, 6085589);
-    public static final MapColor LAPIS = new MapColor(32, 4882687);
-    public static final MapColor EMERALD = new MapColor(33, 55610);
-    public static final MapColor PODZOL = new MapColor(34, 8476209);
-    public static final MapColor NETHER = new MapColor(35, 7340544);
-    public static final MapColor TERRACOTTA_WHITE = new MapColor(36, 13742497);
-    public static final MapColor TERRACOTTA_ORANGE = new MapColor(37, 10441252);
-    public static final MapColor TERRACOTTA_MAGENTA = new MapColor(38, 9787244);
-    public static final MapColor TERRACOTTA_LIGHT_BLUE = new MapColor(39, 7367818);
-    public static final MapColor TERRACOTTA_YELLOW = new MapColor(40, 12223780);
-    public static final MapColor TERRACOTTA_LIGHT_GREEN = new MapColor(41, 6780213);
-    public static final MapColor TERRACOTTA_PINK = new MapColor(42, 10505550);
-    public static final MapColor TERRACOTTA_GRAY = new MapColor(43, 3746083);
-    public static final MapColor TERRACOTTA_LIGHT_GRAY = new MapColor(44, 8874850);
-    public static final MapColor TERRACOTTA_CYAN = new MapColor(45, 5725276);
-    public static final MapColor TERRACOTTA_PURPLE = new MapColor(46, 8014168);
-    public static final MapColor TERRACOTTA_BLUE = new MapColor(47, 4996700);
-    public static final MapColor TERRACOTTA_BROWN = new MapColor(48, 4993571);
-    public static final MapColor TERRACOTTA_GREEN = new MapColor(49, 5001770);
-    public static final MapColor TERRACOTTA_RED = new MapColor(50, 9321518);
-    public static final MapColor TERRACOTTA_BLACK = new MapColor(51, 2430480);
-    public static final MapColor CRIMSON_NYLIUM = new MapColor(52, 12398641);
-    public static final MapColor CRIMSON_STEM = new MapColor(53, 9715553);
-    public static final MapColor CRIMSON_HYPHAE = new MapColor(54, 6035741);
-    public static final MapColor WARPED_NYLIUM = new MapColor(55, 1474182);
-    public static final MapColor WARPED_STEM = new MapColor(56, 3837580);
-    public static final MapColor WARPED_HYPHAE = new MapColor(57, 5647422);
-    public static final MapColor WARPED_WART_BLOCK = new MapColor(58, 1356933);
-    public static final MapColor DEEPSLATE = new MapColor(59, 6579300);
-    public static final MapColor RAW_IRON = new MapColor(60, 14200723);
-    public static final MapColor GLOW_LICHEN = new MapColor(61, 8365974);
+    // Fiddle start - read packs - blocks - get map color by ResourceLocation - add explicit ResourceLocation to MapColor
+    public static final MapColor NONE = new MapColor("none", 0, 0);
+    public static final MapColor GRASS = new MapColor("grass", 1, 8368696);
+    public static final MapColor SAND = new MapColor("sand", 2, 16247203);
+    public static final MapColor WOOL = new MapColor("wool", 3, 13092807);
+    public static final MapColor FIRE = new MapColor("fire", 4, 16711680);
+    public static final MapColor ICE = new MapColor("ice", 5, 10526975);
+    public static final MapColor METAL = new MapColor("metal", 6, 10987431);
+    public static final MapColor PLANT = new MapColor("plant", 7, 31744);
+    public static final MapColor SNOW = new MapColor("snow", 8, 16777215);
+    public static final MapColor CLAY = new MapColor("clay", 9, 10791096);
+    public static final MapColor DIRT = new MapColor("dirt", 10, 9923917);
+    public static final MapColor STONE = new MapColor("stone", 11, 7368816);
+    public static final MapColor WATER = new MapColor("water", 12, 4210943);
+    public static final MapColor WOOD = new MapColor("wood", 13, 9402184);
+    public static final MapColor QUARTZ = new MapColor("quartz", 14, 16776437);
+    public static final MapColor COLOR_ORANGE = new MapColor("color_orange", 15, 14188339);
+    public static final MapColor COLOR_MAGENTA = new MapColor("color_magenta", 16, 11685080);
+    public static final MapColor COLOR_LIGHT_BLUE = new MapColor("color_light_blue", 17, 6724056);
+    public static final MapColor COLOR_YELLOW = new MapColor("color_yellow", 18, 15066419);
+    public static final MapColor COLOR_LIGHT_GREEN = new MapColor("color_light_green", 19, 8375321);
+    public static final MapColor COLOR_PINK = new MapColor("color_pink", 20, 15892389);
+    public static final MapColor COLOR_GRAY = new MapColor("color_gray", 21, 5000268);
+    public static final MapColor COLOR_LIGHT_GRAY = new MapColor("color_light_gray", 22, 10066329);
+    public static final MapColor COLOR_CYAN = new MapColor("color_cyan", 23, 5013401);
+    public static final MapColor COLOR_PURPLE = new MapColor("color_purple", 24, 8339378);
+    public static final MapColor COLOR_BLUE = new MapColor("color_blue", 25, 3361970);
+    public static final MapColor COLOR_BROWN = new MapColor("color_brown", 26, 6704179);
+    public static final MapColor COLOR_GREEN = new MapColor("color_green", 27, 6717235);
+    public static final MapColor COLOR_RED = new MapColor("color_red", 28, 10040115);
+    public static final MapColor COLOR_BLACK = new MapColor("color_black", 29, 1644825);
+    public static final MapColor GOLD = new MapColor("gold", 30, 16445005);
+    public static final MapColor DIAMOND = new MapColor("diamond", 31, 6085589);
+    public static final MapColor LAPIS = new MapColor("lapis", 32, 4882687);
+    public static final MapColor EMERALD = new MapColor("emerald", 33, 55610);
+    public static final MapColor PODZOL = new MapColor("podzol", 34, 8476209);
+    public static final MapColor NETHER = new MapColor("nether", 35, 7340544);
+    public static final MapColor TERRACOTTA_WHITE = new MapColor("terracotta_white", 36, 13742497);
+    public static final MapColor TERRACOTTA_ORANGE = new MapColor("terracotta_orange", 37, 10441252);
+    public static final MapColor TERRACOTTA_MAGENTA = new MapColor("terracotta_magenta", 38, 9787244);
+    public static final MapColor TERRACOTTA_LIGHT_BLUE = new MapColor("terracotta_light_blue", 39, 7367818);
+    public static final MapColor TERRACOTTA_YELLOW = new MapColor("terracotta_yellow", 40, 12223780);
+    public static final MapColor TERRACOTTA_LIGHT_GREEN = new MapColor("terracotta_light_green", 41, 6780213);
+    public static final MapColor TERRACOTTA_PINK = new MapColor("terracotta_pink", 42, 10505550);
+    public static final MapColor TERRACOTTA_GRAY = new MapColor("terracotta_gray", 43, 3746083);
+    public static final MapColor TERRACOTTA_LIGHT_GRAY = new MapColor("terracotta_light_gray", 44, 8874850);
+    public static final MapColor TERRACOTTA_CYAN = new MapColor("terracotta_cyan", 45, 5725276);
+    public static final MapColor TERRACOTTA_PURPLE = new MapColor("terracotta_purple", 46, 8014168);
+    public static final MapColor TERRACOTTA_BLUE = new MapColor("terracotta_blue", 47, 4996700);
+    public static final MapColor TERRACOTTA_BROWN = new MapColor("terracotta_brown", 48, 4993571);
+    public static final MapColor TERRACOTTA_GREEN = new MapColor("terracotta_green", 49, 5001770);
+    public static final MapColor TERRACOTTA_RED = new MapColor("terracotta_red", 50, 9321518);
+    public static final MapColor TERRACOTTA_BLACK = new MapColor("terracotta_black", 51, 2430480);
+    public static final MapColor CRIMSON_NYLIUM = new MapColor("crimson_nylium", 52, 12398641);
+    public static final MapColor CRIMSON_STEM = new MapColor("crimson_stem", 53, 9715553);
+    public static final MapColor CRIMSON_HYPHAE = new MapColor("crimson_hyphae", 54, 6035741);
+    public static final MapColor WARPED_NYLIUM = new MapColor("warped_nylium", 55, 1474182);
+    public static final MapColor WARPED_STEM = new MapColor("warped_stem", 56, 3837580);
+    public static final MapColor WARPED_HYPHAE = new MapColor("warped_hyphae", 57, 5647422);
+    public static final MapColor WARPED_WART_BLOCK = new MapColor("warped_wart_block", 58, 1356933);
+    public static final MapColor DEEPSLATE = new MapColor("deepslate", 59, 6579300);
+    public static final MapColor RAW_IRON = new MapColor("raw_iron", 60, 14200723);
+    public static final MapColor GLOW_LICHEN = new MapColor("glow_lichen", 61, 8365974);
+    public final net.minecraft.resources.ResourceLocation resourceLocation;
+    // Fiddle end - read packs - blocks - get map color by ResourceLocation - add explicit ResourceLocation to MapColor
     public final int col;
     public final int id;
 
-    private MapColor(int id, int color) {
+    // Fiddle start - read packs - blocks - get map color by ResourceLocation - add explicit ResourceLocation to MapColor
+    private MapColor(String name, int id, int color) {
+        this.resourceLocation = new net.minecraft.resources.ResourceLocation(net.minecraft.resources.ResourceLocation.DEFAULT_NAMESPACE, name);
+        // Fiddle end - read packs - blocks - get map color by ResourceLocation - add explicit ResourceLocation to MapColor
         if (id >= 0 && id <= 63) {
             this.id = id;
             this.col = color;
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/read/fiddlecontent/definition/block/mapcolor/MapColorsByResourceLocation.java b/src/main/java/org/fiddlemc/fiddle/pack/read/fiddlecontent/definition/block/mapcolor/MapColorsByResourceLocation.java
new file mode 100644
index 0000000000000000000000000000000000000000..526f4d9df8181ff061434234130afcc0f8989238
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/read/fiddlecontent/definition/block/mapcolor/MapColorsByResourceLocation.java
@@ -0,0 +1,41 @@
+// Fiddle - read packs - blocks - get map color by ResourceLocation
+
+package org.fiddlemc.fiddle.pack.read.fiddlecontent.definition.block.mapcolor;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.level.material.MapColor;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Provides {@link MapColor}s {@linkplain #get by their namespaced key}.
+ * <p>
+ *     The namespaced key of a {@link MapColor} is not something that exists in Minecraft:
+ *     they are not data-driven. However, to support a data-driven approach in configuring for example
+ *     the map color belonging to a block, we explicitly assign a namespaced key to each existing value,
+ *     with the {@link ResourceLocation#DEFAULT_NAMESPACE}.
+ * </p>
+ */
+public final class MapColorsByResourceLocation {
+
+    private MapColorsByResourceLocation() {
+        throw new RuntimeException();
+    }
+
+    private static final Map<ResourceLocation, MapColor> BY_RESOURCE_LOCATION;
+    static {
+        var values = Arrays.stream(MapColor.MATERIAL_COLORS).filter(Objects::nonNull).toArray(MapColor[]::new);
+        BY_RESOURCE_LOCATION = new HashMap<>(values.length);
+        for (var mapColor : values) {
+            BY_RESOURCE_LOCATION.put(mapColor.resourceLocation, mapColor);
+        }
+    }
+
+    public static MapColor get(ResourceLocation resourceLocation) {
+        return BY_RESOURCE_LOCATION.get(resourceLocation);
+    }
+
+}
